### ISSUE TITLE
python's multiprocessing produces inconsistent logs

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#293]
In GitLab by @glennklockwood on Jan 29, 2021, 18:25

Python applications which use `multiprocessing` result in Darshan logs with really strange outputs; I think this is a result of multiprocessing using clone(2) with `child_stack=0`, forcing copy-on-write semantics for the process stack.

Reproducer:

```python
import os
import multiprocessing

def f(x):
    fname = 'output.%d' % x
    with open(fname, 'w') as fp:
        fp.write("my pid is %d" % os.getpid())
    return fname

if __name__ == '__main__':
    print("I am the parent process and my pid is %d" % os.getpid())
    with multiprocessing.Pool(5) as p:
        fnames = p.map(f, range(10))
    for fname in fnames:
        print("Created %s with contents: %s" % (fname, open(fname, 'r').read()))
```

When run,

```
$ LD_PRELOAD="/global/homes/g/glock/apps.cori-haswell/darshan.forksafe/lib/libdarshan.so" DARSHAN_ENABLE_NONMPI=1 python test_multiprocessing.py
I am the parent process and my pid is 9617
Created output.0 with contents: my pid is 9618
Created output.1 with contents: my pid is 9619
Created output.2 with contents: my pid is 9620
Created output.3 with contents: my pid is 9621
Created output.4 with contents: my pid is 9622
Created output.5 with contents: my pid is 9618
Created output.6 with contents: my pid is 9619
Created output.7 with contents: my pid is 9620
Created output.8 with contents: my pid is 9621
Created output.9 with contents: my pid is 9622
```

Five child processes are created, each writing to two files.  The parent process then serially opens and reads the contents of those files.  However the Darshan log seems pretty wild:

```
$ darshan-parser glock_python_id*.darshan | egrep -o 'POSIX_(READS|WRITES).*output..'
POSIX_READS	2	/global/u2/g/glock/output.0
POSIX_WRITES	5	/global/u2/g/glock/output.0
POSIX_READS	2	/global/u2/g/glock/output.1
POSIX_WRITES	5	/global/u2/g/glock/output.1
POSIX_READS	2	/global/u2/g/glock/output.2
POSIX_WRITES	0	/global/u2/g/glock/output.2
POSIX_READS	2	/global/u2/g/glock/output.3
POSIX_WRITES	0	/global/u2/g/glock/output.3
POSIX_READS	2	/global/u2/g/glock/output.4
POSIX_WRITES	0	/global/u2/g/glock/output.4
POSIX_READS	2	/global/u2/g/glock/output.5
POSIX_WRITES	0	/global/u2/g/glock/output.5
POSIX_READS	2	/global/u2/g/glock/output.6
POSIX_WRITES	0	/global/u2/g/glock/output.6
POSIX_READS	2	/global/u2/g/glock/output.7
POSIX_WRITES	0	/global/u2/g/glock/output.7
POSIX_READS	2	/global/u2/g/glock/output.8
POSIX_WRITES	0	/global/u2/g/glock/output.8
POSIX_READS	2	/global/u2/g/glock/output.9
POSIX_WRITES	0	/global/u2/g/glock/output.9
```

All the writes from the child processes are missing except for two files whose writes appear over-represented.

It looks like the constructor/destructor hooks we use function with clone(2), but the children are somehow able to affect the memory space of the parent (or something).  I don't have a good enough grasp of the semantics of clone to devise a workaround.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Feb 4, 2021, 17:15

I've been trying to tinker around with this, and I get different behavior (although still not correct behavior):

```
shane@shane-x1-carbon ~/software/darshan/darshan-dev/darshan-runtime/mp-tests (dev-fork-safe) $ darshan-parser ~/software/darshan/darshan-logs/2021/2/4/shane_python_id128589-128589_2-4-60913-7035573431850780836_1612479314.darshan | egrep -o 'STDIO_(READS|WRITES).*output..'
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.0
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.0
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.1
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.1
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.2
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.2
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.3
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.3
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.4
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.4
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.5
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.5
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.6
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.6
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.7
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.7
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.8
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.8
STDIO_READS	1	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.9
STDIO_WRITES	0	/home/shane/software/darshan/darshan-dev/darshan-runtime/mp-tests/output.9

```

Two things:
1. I don't see any read/write activity at the POSIX layer according to Darshan.
2. I'm only able to instrument the STDIO read calls on the parent process, the log contains no write information (STDIO or POSIX).

I'm still trying to think through this, but one quick question is whether you are using `--enable-mmap-logs`? The way the mmap logs are setup, child processes inherit the mmap memory and any updates they make should be carried through to the parent memory (we map with MAP_SHARED). If you are using this combination of features, maybe it leads to nonsensical counters like that? In general, we'll probably want to avoid using the mmap logs with non-MPI multi-process (i.e, fork/exec/clone) uses cases.

If you aren't using mmap logs, then copy-on-write semantics should make it s.t. children can not modify parent memory, so something else could be happening.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Feb 4, 2021, 20:28

(Ignoring the mmap issue for now)

I was able to add some printf statements and confirm that the Darshan shared library destructor does _not_ run on cloned processes (it does on the parent) -- that means the cloned processes don't get a chance to shutdown Darshan and write out a log. This is different than fork, where children and parents both call destructors and write out Darshan logs. I can't find any language in manpages or elsewhere that explains this, but it is what I see in practice -- this is probably just one of the subtle differences between clone() and fork() functions.

Even if destructors were called on every process, cloned processes would still contain a mix of pre-clone parent statistics and post-clone child statistics, which is probably not that useful. `pthread_atfork()` allows us to account for this in the case of applications calling fork (child processes reset Darshan runtime state to get a clean slate), but it does not work with clone (and I don't see any sort of equivalent functionality).

Doesn't look promising that Darshan can be made to work in this case.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on Feb 8, 2021, 13:58

OK, maybe this is just a notice that gets buried in the documentation somewhere.  Or maybe a flag raised in the log if clone is detected--unsure how many ways there are to issue the syscall from C though.

The mmap thing did occur to me and I ran with+without it and didn't get sane results either way, but I can't find the results of that test and it doesn't sound like it would make much of a difference if applications using clone are simply unsupported.  Thanks for looking into this.

**Comment by shanedsnyder** [comment]
We made some recent progress on getting Darshan to properly instrument Python `multiprocessing` in #978, so wanted to revisit this hoping to have resolution and sane looking logs for this case.

Alas, Python had other plans (and somewhat dubious best practices for doing multiprocessing)...

Turns out the `Pool` context manager here is essentially resulting in the subprocesses being killed (with `SIGTERM`) when releasing resources:
```
    with multiprocessing.Pool(5) as p:
        fnames = p.map(f, range(10))
```
That leaves Darshan no shutdown hooks (either using it's shared library destructor or the `_exit()` wrapper introduced in the above PR) that it can use to write out it's log file. There's honestly not much Darshan can hope to do on the subprocesses in that case -- writing out a log file is not something that is safe to do in a signal handler, AFAIK.

So, ultimately, I don't think this is something we'll be able to support in Darshan. I'll go ahead and close this issue for now.

FWIW, I do get expected Darshan output on all processes when I'm extra careful about how the subprocesses are terminating:
```
    with multiprocessing.Pool(5) as p:
        fnames = p.map(f, range(10))
        p.close()
        p.join()
```
